xls/modules/dbe/lz4_decoder.x

```diff
@@ -0,0 +1,618 @@
+    // generate an error, as otherwise that will make decoder output garbage
+    // data that may leak some stale data from previous blocks)
+    hb_all_valid: bool,
+}
+
+pub proc decoder_base<SYM_WIDTH: u32, PTR_WIDTH: u32, CNT_WIDTH: u32> {
+    i_encoded: chan<Token<SYM_WIDTH, PTR_WIDTH, CNT_WIDTH>> in;
+    o_data: chan<PlainData<SYM_WIDTH>> out;
+
+    /// History buffer RAM
+    /// Size: (1<<PTR_WIDTH) x SYM_WIDTH
+    o_ram_hb_wr_req: chan<ram::WriteReq<PTR_WIDTH, SYM_WIDTH, 1>> out;
+    i_ram_hb_wr_resp: chan<ram::WriteResp> in;
+    o_ram_hb_rd_req: chan<ram::ReadReq<PTR_WIDTH, 1>> out;
+    i_ram_hb_rd_resp: chan<ram::ReadResp<SYM_WIDTH>> in;
+
+    init {
+        State<SYM_WIDTH, PTR_WIDTH, CNT_WIDTH> {
+            wr_ptr: uN[PTR_WIDTH]:0,
+            rd_ptr: uN[PTR_WIDTH]:0,
+            rd_ctr_rem: uN[CNT_WIDTH]:0,
+            err: false,
+            hb_all_valid: false,
+        }
+    }
+
+    config (
+        i_encoded: chan<Token<SYM_WIDTH, PTR_WIDTH, CNT_WIDTH>> in,
+        o_symb: chan<PlainData<SYM_WIDTH>> out,
+        // RAM
+        o_ram_hb_wr_req: chan<ram::WriteReq<PTR_WIDTH, SYM_WIDTH, 1>> out,
+        i_ram_hb_wr_resp: chan<ram::WriteResp> in,
+        o_ram_hb_rd_req: chan<ram::ReadReq<PTR_WIDTH, 1>> out,
+        i_ram_hb_rd_resp: chan<ram::ReadResp<SYM_WIDTH>> in,
+    ) {
+        (
+            i_encoded, o_symb,
+            o_ram_hb_wr_req, i_ram_hb_wr_resp,
+            o_ram_hb_rd_req, i_ram_hb_rd_resp,
+        )
+    }
+
+    next (tok: token, state: State<SYM_WIDTH, PTR_WIDTH, CNT_WIDTH>) {
+        let is_copying = state.rd_ctr_rem != uN[CNT_WIDTH]:0;
+        let do_recv = !is_copying;
+
+        // Receive token
+        let (tok, rx) = recv_if(tok, i_encoded, do_recv,
+            dbe::zero_token<SYM_WIDTH, PTR_WIDTH, CNT_WIDTH>());
+
+        let rx_lt = do_recv && rx.kind == TokenKind::LT;
+        let rx_cp = do_recv && rx.kind == TokenKind::CP;
+        let rx_mark = do_recv && rx.kind == TokenKind::MK;
+        let rx_end = rx_mark && rx.mark == Mark::END;
+        let rx_error = rx_mark && dbe::is_error(rx.mark);
+        let rx_reset = rx_mark && rx.mark == Mark::RESET;
+        let rx_unexpected_mark = rx_mark && !(rx_end || rx_reset || rx_error);
+
+        // Are we going to output a literal symbol or a symbol from HB?
+        let emit_sym_lit = rx_lt;
+        let emit_sym_from_hb = rx_cp || is_copying;
+
+        // Reset, set or increment read pointer
+        let rd_ptr = if rx_reset {
+            uN[PTR_WIDTH]:0
+        } else if rx_cp {
+            state.wr_ptr - rx.cp_off - uN[PTR_WIDTH]:1
+        } else if is_copying {
+            state.rd_ptr + uN[PTR_WIDTH]:1
+        } else {
+            state.rd_ptr
+        };
+
+        // Check whether rd_ptr points to an unwritten HB cell
+        let rd_ptr_valid = (rd_ptr < state.wr_ptr) || state.hb_all_valid;
+
+        // HB READ - read the symbol from history buffer
```

Comment:
Does this handle the "overlap match" scenario referenced in https://github.com/lz4/lz4/blob/dev/doc/lz4_Block_format.md when the symbol from the history buffer depends on the match copy?

Reply:
I'd leave this to @proppy to decide whether or not the DSLX tests should be more comprehensive.